klippy/extras/display/st7789v.py

```diff
@@ -0,0 +1,531 @@
+    except ValueError:
+        raise ST7789vParseException("Malformed rectangle '%s'"
+                                    % (rect_string,))
+
+    if len(elements) != 4:
+        raise ST7789vParseException("Malformed rectangle '%s'"
+                                    % (rect_string,))
+
+    return {
+        'left': elements[0],
+        'top' : elements[1],
+        'right' : elements[2],
+        'bottom' : elements[3]}
+
+def _parse_color(color_string):
+    components = color_string.split(',')
+
+    try:
+        elements = tuple(max(0, min(255, math.floor(float(p.strip()))))
+                         for p in components)
+    except ValueError:
+        raise ST7789vParseException("Malformed color '%s'" % (color_string,))
+
+    if len(elements) != 3:
+        raise ST7789vParseException("Malformed color '%s'" % (color_string,))
+
+    # convert to RGB565
+    _r = int(elements[0])
+    _g = int(elements[1])
+    _b = int(elements[2])
+
+    color = ((_r // 8) << 11) | ((_g // 4) << 5) | (_b // 8)
+    return color
+
+def _parse_glyph(data):
+    glyph_data = []
+    width = None
+    height = 0
+
+    for line in data.split('\n'):
+        line = line.strip().replace('.', '0').replace('*', '1')
+        if not line:
+            continue
+        height = height + 1
+        if width is None:
+            width = len(line)
+        elif len(line) != width or line.replace('0', '').replace('1', ''):
+            raise ST7789vParseException("Invalid glyph line %d" % (height,))
+
+        linedata = bytearray((width+7) // 8)
+        index = 0
+        while line:
+            this_chunk = line[0:8]
+            line = line[8:]
+
+            # pad chunk if less than 8 pixels wide
+            chunk_length = len(this_chunk)
+            if chunk_length != 8:
+                this_chunk += '0' * (8 - chunk_length)
+
+            linedata[index] = int(this_chunk, 2)
+            index += 1
+
+        glyph_data.append(linedata)
+
+    return glyph_data, width, height
+
+class ResetHelper(object):
+    def __init__(self, disp, resx_pin, csx_pin, rdx_pin, mcu, cmd_queue):
+        self.mcu_csx = None
+        self.mcu_rdx = None
+        self.display = disp
+
+        self.mcu = mcu
+        self.cmd_queue = cmd_queue
+
+        self.mcu_resx = bus.MCU_bus_digital_out(
+            self.mcu,
+            resx_pin,
+            self.cmd_queue)
+
+        if csx_pin is not None:
+            self.mcu_csx = bus.MCU_bus_digital_out(
+                self.mcu,
+                csx_pin,
+                self.cmd_queue)
+
+        if rdx_pin is not None:
+            self.mcu_rdx = bus.MCU_bus_digital_out(
+                self.mcu,
+                rdx_pin,
+                self.cmd_queue)
+
+    def init(self):
+        curtime = self.mcu.get_printer().get_reactor().monotonic()
+
+        print_time = self.mcu.estimated_print_time(curtime)
+        print_time += ST7789_RESET_PULSE_LENGTH
```

Comment:
The `minclock` system of timing isn't accurate to the millisecond.  The `minclock` system is implemented by timing release of messages from the host serial buffers, and because of retransmits, buffering delays, etc, it isn't that accurate.  So, if you want to time updates to guarantee a 5 millisecond delay, then it's necessary to "hit it with a hammer" and delay by 100ms or so.

Reply:
Ah, ok. I implemented it like this based on your advice from last july -i've been sitting on this PR for a while-, but missed the reduced accuracy. 

The reset pulse needs to be >10us, so waiting 100 ms seems a bit excessive, is there a middle ground?

Reply:
Maybe.  It's hard to put a limit on the serial timing though.  If there is only a couple of sync points and it is only done at startup, then it probably wont matter if one uses 100ms.